cirq/experiments/t2_decay_experiment.py

```diff
@@ -0,0 +1,336 @@
+                                 start=min_delay_dur.total_nanos(),
+                                 stop=max_delay_dur.total_nanos(),
+                                 length=num_points)
+
+    if experiment_type == ExperimentType.RAMSEY:
+        # Ramsey T2* experiment
+        # Use sqrt(Y) to flip to the equator.
+        # Evolve the state for a given amount of delay time
+        # Then measure the state in both X and Y bases.
+
+        circuit = circuits.Circuit(
+            ops.Y(qubit)**0.5,
+            ops.WaitGate(value.Duration(nanos=delay_var))(qubit),
+            ops.X(qubit)**inv_x_var,
+            ops.Y(qubit)**inv_y_var,
+            ops.measure(qubit, key='output'),
+        )
+        tomography_sweep = study.Zip(
+            study.Points('inv_x', [0.0, -0.5]),
+            study.Points('inv_y', [-0.5, 0.0]),
+        )
+        sweep = study.Product(delay_sweep, tomography_sweep)
+    elif experiment_type == ExperimentType.HAHN_ECHO:
+        # Hahn / Spin Echo T2 experiment
+        # Use sqrt(Y) to flip to the equator.
+        # Evolve the state for half the given amount of delay time
+        # Flip the state using an X gate
+        # Evolve the state for half the given amount of delay time
+        # Then measure the state in both X and Y bases.
+
+        # TODO: Should be 0.5 * delay_var
+        circuit = circuits.Circuit(
+            ops.Y(qubit)**0.5,
+            ops.WaitGate(value.Duration(nanos=delay_var))(qubit),
+            ops.X(qubit),
+            ops.WaitGate(value.Duration(nanos=delay_var))(qubit),
+            ops.X(qubit)**inv_x_var,
+            ops.Y(qubit)**inv_y_var,
+            ops.measure(qubit, key='output'),
+        )
+        tomography_sweep = study.Zip(
+            study.Points('inv_x', [0.0, 0.5]),
+            study.Points('inv_y', [-0.5, 0.0]),
+        )
+        sweep = study.Product(delay_sweep, tomography_sweep)
+    else:
+        raise ValueError(f'Experiment type {experiment_type} not supported')
+
+    # Tabulate measurements into a histogram
+    results = sampler.sample(circuit, params=sweep, repetitions=repetitions)
+
+    y_basis_measurements = results[results.inv_y < 0]
+    y_basis_tabulation = pd.crosstab(y_basis_measurements.delay_ns,
+                                     y_basis_measurements.output).reset_index()
+    x_basis_measurements = results[abs(results.inv_x) > 0]
+    x_basis_tabulation = pd.crosstab(x_basis_measurements.delay_ns,
+                                     x_basis_measurements.output).reset_index()
+
+    # If all measurements are 1 or 0, fill in the missing column with all zeros.
+    for tab in [x_basis_tabulation, y_basis_tabulation]:
+        for col_index, name in [(1, 0), (2, 1)]:
+            if name not in tab:
+                tab.insert(col_index, name, [0] * tab.shape[0])
+
+    # Return the results in a container object
+    return T2DecayResult(x_basis_tabulation, y_basis_tabulation)
+
+
+class T2DecayResult:
```

Comment:
This is cool! Are there plans for some kind of abstract class that provides expectations with respect to specific runs/sweeps? This seems like a special case of that

Reply:
@mpharrigan is working on a framework to do something like this.  Matt, are you presenting your proposed observation / expectation idea during any of the upcoming cirq cyncs?  (you should).

Reply:
It's been pushed to the back burner because of some other projects, but I should have some code to share soon!

Reply:
Going over the design doc with the public cirq community might be good to get feedback before you do too much coding though!

Reply:
cool, looking forward to it